homeassistant/components/homekit/config_flow.py

```diff
@@ -278,46 +286,78 @@ async def async_step_cameras(self, user_input=None):
-                ),
-                CONF_INCLUDE_ENTITIES: self.homekit_options.get(
-                    CONF_INCLUDE_ENTITIES, []
-                ),
-                CONF_EXCLUDE_ENTITIES: user_input[CONF_EXCLUDE_ENTITIES],
+            entity_filter = {
+                CONF_INCLUDE_DOMAINS: [],
+                CONF_EXCLUDE_DOMAINS: [],
+                CONF_INCLUDE_ENTITIES: [],
+                CONF_EXCLUDE_ENTITIES: [],
             }
-            for entity_id in user_input[CONF_EXCLUDE_ENTITIES]:
-                if entity_id in self.included_cameras:
-                    self.included_cameras.remove(entity_id)
+
+            if user_input[CONF_INCLUDE_EXCLUDE_MODE] == MODE_INCLUDE:
+                entity_filter[CONF_INCLUDE_ENTITIES] = user_input[CONF_ENTITIES]
+                # Include all of the domain if there are no entities
+                # explicitly included as the user selected the domain
+                domains_with_entities_selected = _domains_set_from_entities(
+                    user_input[CONF_ENTITIES]
+                )
+                entity_filter[CONF_INCLUDE_DOMAINS] = [
+                    domain
+                    for domain in self.homekit_options[CONF_DOMAINS]
+                    if domain not in domains_with_entities_selected
+                ]
+
+                for entity_id in list(self.included_cameras):
+                    if entity_id not in user_input[CONF_ENTITIES]:
+                        self.included_cameras.remove(entity_id)
+            else:
+                entity_filter[CONF_INCLUDE_DOMAINS] = self.homekit_options[CONF_DOMAINS]
+                entity_filter[CONF_EXCLUDE_ENTITIES] = user_input[CONF_ENTITIES]
+                for entity_id in user_input[CONF_ENTITIES]:
+                    if entity_id in self.included_cameras:
+                        self.included_cameras.remove(entity_id)
+
+            self.homekit_options[CONF_FILTER] = entity_filter
+
             if self.included_cameras:
                 return await self.async_step_cameras()
+
             return await self.async_step_advanced()
 
         entity_filter = self.homekit_options.get(CONF_FILTER, {})
         all_supported_entities = await self.hass.async_add_executor_job(
             _get_entities_matching_domains,
             self.hass,
-            self.homekit_options[CONF_INCLUDE_DOMAINS],
+            self.homekit_options[CONF_DOMAINS],
         )
         self.included_cameras = {
             entity_id
             for entity_id in all_supported_entities
             if entity_id.startswith("camera.")
         }
+
+        entities = entity_filter.get(CONF_INCLUDE_ENTITIES, [])
+        if entities:
+            include_exclude_mode = MODE_INCLUDE
+        else:
+            include_exclude_mode = MODE_EXCLUDE
+            entities = entity_filter.get(CONF_EXCLUDE_ENTITIES, [])
+
         data_schema = vol.Schema(
             {
+                vol.Required(
```

Comment:
Instead of doing all this, I wonder if we should add a new type "specify entity include/excludes" and have it render in the frontend with a native widget. We already define `cv.custom_serializer` and could add support for the entity filter schema on the backend.

@bramkragten what do you think?

Reply:
This is an idea for a future PR btw, it shouldn't block this.